### ISSUE TITLE
JsonConfigFileAccessor Needs Inter-Process Synchronization to handle Competing IO

### ISSUE THREAD

**Issue by mirichmo** [PowerShell/PowerShell#1630]
JsonConfigFileProcessor needs inter-process synchronization to handle competing IO from different processes on the same file. If a process attempts to read the file while another process has the write lock, the read will fail. Within a single process, this is gracefully managed by ReaderWriterLockSlim. We need an inter-process version of it for this scenario.
## Steps to reproduce
1. Start powershell and attach a debugger to it. 
2. Set a BP in SetExecutionPolicy
3. `Set-ExecutionPolicy -Scope LocalMachine Restricted`
4. Step through the code until it opens the FileStream for the config file
5. In another process, start PowerShell and `Get-ExecutionPolicy -list`
## Expected behavior

The second process should block until the first finishes its write.
## Actual behavior

``` powershell
PS > get-executionpolicy -list

        Scope ExecutionPolicy
        ----- ---------------
MachinePolicy       Undefined
   UserPolicy       Undefined
      Process       Undefined
get-executionpolicy : The process cannot access the file
'C:\Users\<user>\AppData\Local\PowerShell\Configuration\PowerShellProperties.json' because it is being used by another process.
At line:1 char:1
+ get-executionpolicy -list
+ ~~~~~~~~~~~~~~~~~~~~~~~~~
    + CategoryInfo          : NotSpecified: (:) [Get-ExecutionPolicy], IOException
    + FullyQualifiedErrorId : System.IO.IOException,Microsoft.PowerShell.Commands.GetExecutionPolicyCommand
```
## Environment data

``` powershell
Name                           Value
----                           -----
GitCommitId                    v6.0.0-alpha.7-116-gd26b614837f657c1984d0f5e89c9395e4c56e6b8
SerializationVersion           1.1.0.1
PSRemotingProtocolVersion      2.3
CLRVersion
PSEdition                      PowerShellCore
WSManStackVersion              3.0
BuildVersion                   3.0.0.0
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0...}
PSVersion                      6.0.0-alpha
```

**Comment by joeyaiello** [comment]
@mirichmo this looks like a pretty rare edge case. How would a user reasonably hit/be affected by this problem? Is this something that you see as necessary 6.0?

**Comment by mirichmo** [comment]
At the moment, it is not highly likely that users will encounter it, but its chances of occurring increase with each additional value that we move to the configuration file. It is a real issue (as demonstrated above) and should not be closed, but it can be moved to 6.1 if you are looking to reduce the number of 6.0 issues.
